### ISSUE TITLE
Semantic Segmentation Experiment: Predicted labels are correct in `test_preds.png` but the actual predictions are wrong

### ISSUE THREAD

**Issue by theoway** [azavea/raster-vision#1549]
## To Reproduce

1. Use docker image rastervision-pytorch-v.0.13.1
1. Use the model_bundle.zip and rionegro_rgb_8_2.tif for `predict`. Files are available [here](https://drive.google.com/drive/folders/18kpE3y51LYlYzyborrE6oojNC4oFaSAt?usp=share_link).
1. This model was trained for prediction of cars and motorcyles. You should be getting correct predictions, but won't happen. 
You will get incorrect results similar to `labels.tif` file that is [here](https://drive.google.com/drive/folders/18kpE3y51LYlYzyborrE6oojNC4oFaSAt?usp=share_link).
Also, there should be vector outputs as well in the output folder that you provide. But that doesn't happen either.

The model had been trained using AWS batch. Here's the `get_config()` code:
```python
import os

from rastervision.core.rv_pipeline import *
from rastervision.core.backend import *
from rastervision.core.data import *
from rastervision.pytorch_backend import *
from rastervision.pytorch_learner import *

IDS = ['El_Retiro_1_3', 'El_Retiro_1_4', 'El_Retiro_2_1', ...]

def build_scene(uri, id, channel_order=None):
    label_file = id + '.geojson'
    label_uri = uri + label_file
    print("Label uri: ", label_uri)

    image_file = id + '.tif'
    image_uri = uri + image_file
    print("Image uri: ", image_uri)

    raster_source = RasterioSourceConfig(
        uris=[image_uri],
        channel_order=channel_order,
        transformers=[StatsTransformerConfig()])
    
    vector_source = GeoJSONVectorSourceConfig(
        uri=label_uri,
        default_class_id=0,
        ignore_crs_field=True)

    label_source = SemanticSegmentationLabelSourceConfig(
        raster_source=RasterizedSourceConfig(
            vector_source=vector_source,
            rasterizer_config=RasterizerConfig(background_class_id=0)))
    
    label_store = SemanticSegmentationLabelStoreConfig(
        vector_output=[PolygonVectorOutputConfig(class_id=1),
                        PolygonVectorOutputConfig(class_id=2),
                        PolygonVectorOutputConfig(class_id=3)])

    return SceneConfig(
        id=id,
        raster_source=raster_source,
        label_source=label_source,
        label_store=label_store)

def get_config(runner, raw_uri, root_uri, test=True):
    scene_ids = IDS

    split_ratio = 0.8
    num_train_ids = round(len(scene_ids) * split_ratio)
    train_ids = scene_ids[0:num_train_ids]
    val_ids = scene_ids[num_train_ids:]

    num_epochs = 200

    if test:
        train_ids = train_ids[:5:]
        val_ids = val_ids[:3:]

        num_epochs = 2
    
    class_config = ClassConfig(
        names=['background', 'motorcycle', 'car', 'ghost'],
        colors=['#ffff00', '#0000ff', '#00ffff', '#00ff00'],
        null_class='background'
    )

    channel_order = [0, 1, 2]
    train_scenes = [
        build_scene(raw_uri, id, channel_order) for id in train_ids
    ]
    val_scenes = [
        build_scene(raw_uri, id, channel_order) for id in val_ids
    ]

    dataset = DatasetConfig(
        class_config=class_config,
        train_scenes=train_scenes,
        validation_scenes=val_scenes)
    
    chip_sz = 325
    img_sz = chip_sz
    chip_options = SemanticSegmentationChipOptions(
        window_method=SemanticSegmentationWindowMethod.sliding,
        stride=img_sz)

    data = SemanticSegmentationImageDataConfig(
            img_sz=img_sz, num_workers=4)

    backend = PyTorchSemanticSegmentationConfig(
        data=data,
        model=SemanticSegmentationModelConfig(backbone=Backbone.resnet50),
        solver=SolverConfig(
            lr=1e-4,
            num_epochs=num_epochs,
            test_num_epochs=2,
            batch_sz=8,
            one_cycle=True))

    return SemanticSegmentationConfig(
        root_uri=root_uri,
        dataset=dataset,
        backend=backend,
        train_chip_sz=img_sz,
        predict_chip_sz=img_sz,
        chip_options=chip_options)
```

## Expected behavior

Please have a look at `eval.json` and `test_preds.jpeg`. Files available [here](https://drive.google.com/drive/folders/18kpE3y51LYlYzyborrE6oojNC4oFaSAt?usp=share_link).
According to them, the model has been trained properly as can be seen in the predicted labels at `test_preds.jpeg`.
Moreover, it should also provide vector/geojson of each class, as I have already configured the `get_config()` code that way. If you see, I've already added:
```python
    label_store = SemanticSegmentationLabelStoreConfig(
        vector_output=[PolygonVectorOutputConfig(class_id=1),
                        PolygonVectorOutputConfig(class_id=2),
                        PolygonVectorOutputConfig(class_id=3)])

    return SceneConfig(
        id=id,
        raster_source=raster_source,
        label_source=label_source,
        label_store=label_store)
```

## Environment

Docker image: **/pytorch-v.0.13.1**

**Comment by theoway** [comment]
I have spent $20 on the training, using AWS batch. I'm really hoping all the effort doesn't go to waste 😥
Ping me anytime if you need more information.

**Comment by AdeelH** [comment]
The inputs in `test_preds.png` seem to have everything other than cars blacked out, but you test image (`rionegro_rgb_8_2.tif`) does not. Could this be the issue?

**Comment by theoway** [comment]
> The inputs in `test_preds.png` seem to have everything other than cars blacked out, but your test image (`rionegro_rgb_8_2.tif`) does not. Could this be the issue?

This makes sense... I'll try `predict` command with an input image with background already blacked out. Will share the results.

**Comment by theoway** [comment]
@AdeelH In the meanwhile, can you please tell me how should I configure my `get_config()` python file? My goals is to use a trained model to predict cars/motorcycles etc. whose labels I have. 
Ideally, I should only be getting these labels when I provide an image like this, `rionegro_rgb_8_2.tif`, not having to masking the background prior, as it'd defeat the purpose of having an ML model.

I feel I should not have added `background` in class config, as this class is not labelled. The geojson files only have three classes:
```
{
   "type":"FeatureCollection",
   "name":"C:\\Users\\john\\Downloads\\AutoMap\\HAGDAVS\\BBs\\Rionegro_BBs_2_1.geojson",
   "crs":{
      "type":"name",
      "properties":{
         "name":"urn:ogc:def:crs:OGC:1.3:CRS84"
      }
   },
   "features":[
      {
         "type":"Feature",
         "properties":{
            "OBJECTID":"1047",
            "Class":"3",
            "Type":"",
            "Shape_Leng":"7.72813431718e-005",
            "Shape_Area":"3.71918079612e-010"
         },
         "geometry":{
            "type":"Polygon",
            "coordinates":[
               [
                  [
                     -75.380883500073423,
                     6.147653802281184
                  ],
                  [
                     -75.380886276280592,
                     6.147648052016052
                  ],
                  [
                     -75.3809025360232,
                     6.147651026973392
                  ],
                  [
                     -75.380902338172348,
                     6.147672441629936
                  ],
                  [
                     -75.380881715818475,
                     6.147664510508775
                  ],
                  [
                     -75.380883500073423,
                     6.147653802281184
                  ]
               ]
            ]
         }
      },
      {
         "type":"Feature",
         "properties":{
            "OBJECTID":"1048",
            "Class":"1",
            "Type":"",
            "Shape_Leng":"6.60527234097e-005",
            "Shape_Area":"1.54425486698e-010"
         },
         "geometry":{
            "type":"Polygon",
            "coordinates":[
               [
                  [
                     -75.380871999543103,
                     6.147651026973392
                  ],
                  [
                     -75.380861886666708,
                     6.147647457564176
                  ],
                  [
                     -75.380854352146571,
                     6.147647457564176
                  ],
                  [
                     -75.380852369141508,
                     6.147644086905132
                  ],
                  [
                     -75.380846222275295,
                     6.147642698351888
                  ],
                  [
                     -75.380846222275295,
                     6.147640715346768
                  ],
                  [
                     -75.380860696863635,
                     6.147638931091819
                  ],
                  [
                     -75.380865654826096,
                     6.147643690304108
                  ],
                  [
                     -75.380874180399076,
                     6.147646267761104
                  ],
                  [
                     -75.380871999543103,
                     6.147651026973392
                  ]
               ]
            ]
         }
      },
      {
         "type":"Feature",
         "properties":{
            "OBJECTID":"1049",
            "Class":"2",
            "Type":"",
            "Shape_Leng":"1.53005995622e-004",
            "Shape_Area":"1.15309196041e-009"
         },
         "geometry":{
            "type":"Polygon",
            "coordinates":[
               [
                  [
                     -75.380734585447726,
                     6.14761609735669
                  ],
                  [
                     -75.380747890058103,
                     6.14761609735669
                  ],
                  [
                     -75.380747831047529,
                     6.147616286162702
                  ],
                  [
                     -75.380734585447726,
                     6.14761609735669
                  ]
               ]
            ]
         }
      }
   ]
}
```

Please guide me on what changes should I make to achieve the desired results 🙏

**Comment by AdeelH** [comment]
>I feel I should not have added background in class config, as this class is not labelled. The geojson files only have three classes:(Class == 1(motorcycle) or 2(car) or 3(ghost))

Your `ClassConfig` is fine. No need to change anything there. Note how you are passing `background_class_id=0` to `RasterizerConfig`. That means "all pixels that don't fall in any polygon in the GeoJSON file should be assigned class 0", which is exactly what you want.

>I should only be getting these labels when I provide an image like this, rionegro_rgb_8_2.tif, not having to masking the background prior

You simply need to train on images that have not had their backgrounds removed.

**Comment by theoway** [comment]
@AdeelH Thank you for the insights! 
But this is weird, my training input images have their backgrounds preserved. I don't know why rastervision masked background of each image when training. Please have a look at this:
![Screenshot (245)](https://user-images.githubusercontent.com/41969735/201835083-35778a9f-452a-4868-b7bf-696f3589c368.png)

Is there any parameter that I have to pass to the pipeline to tell it _not to remove/mask background_ so that training can be done as intended?

Again, I'm learning a lot from you and very grateful for your tips! 💯

**Comment by AdeelH** [comment]
I think I know what is happening: [this call](https://github.com/azavea/raster-vision/blob/3b5ed2373241aed3b1b4595fde902d011214adf6/rastervision_core/rastervision/core/rv_pipeline/semantic_segmentation.py#L140) to [this function](https://github.com/azavea/raster-vision/blob/b6b91fb331cae2d83886f795a87cc30e333af551/rastervision_core/rastervision/core/rv_pipeline/utils.py#L15-L21) is the culprit. 

I do not know why this was added in the first place. It looks wrong to me. Especially since `RasterioSource` fills in NODATA pixels with zeros already. It also only happens if you use `ImageDataConfig` and not `GeoDataConfig` which is probably why I've never run into this before.

Thanks for catching this.

To workaround this, you can just use a separate null class:
```python
    class_config = ClassConfig(
        names=['background', 'motorcycle', 'car', 'ghost'],
        colors=['#ffff00', '#0000ff', '#00ffff', '#00ff00'],
    )
    class_config.ensure_null_class()
```

Also, please do smaller test runs locally for sanity-checking before doing a costly full training run.

**Comment by theoway** [comment]
@AdeelH Thank you so much! 
Should I leave `rasterizer_config=RasterizerConfig(background_class_id=0)))` as it is?

**Comment by AdeelH** [comment]
> Should I leave `rasterizer_config=RasterizerConfig(background_class_id=0)))` as it is?

Yep

**Comment by AdeelH** [comment]
Closing this for now under the assumption that #1556 has fixed this bug. Please reopen if the issue is not resolved.
